enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/pipes/VarLengthExpandRegisterPipe.scala

```diff
@@ -0,0 +1,135 @@
+                                       pipeline: PipelineInformation)
+                                      (val id: Id = new Id) extends PipeWithSource(source) {
+  /*  Since Long is used for both edges and nodes, the "why" of these aliases is to make the code a easier to read.*/
+
+  type LNode = Long
+  type LRelationship = Long
+
+  private def varLengthExpand(node: LNode,
+                              state: QueryState,
+                              row: ExecutionContext): Iterator[(LNode, Seq[LRelationship])] = {
+    val stack = new mutable.Stack[(LNode, Seq[LRelationship])]
+    stack.push((node, Seq.empty))
+
+    new Iterator[(LNode, Seq[LRelationship])] {
+      override def next(): (LNode, Seq[LRelationship]) = {
+        val (fromNode, rels) = stack.pop()
+        if (rels.length < maxDepth.getOrElse(Int.MaxValue) && filteringStep.filterNode(row, state)(fromNode)) {
+          val relationships: RelationshipIterator = state.query.getRelationshipsForIdsPrimitive(fromNode, dir, types.types(state.query))
+
+          var otherSide = 0l
+
+          val relVisitor = new RelationshipVisitor[InternalException] {
+            override def visit(relationshipId: LRelationship, typeId: Int, startNodeId: LNode, endNodeId: LNode): Unit =
+              if (fromNode == startNodeId)
+                otherSide = endNodeId
+              else
+                otherSide = startNodeId
+          }
+
+
+          while (relationships.hasNext) {
+            val relId = relationships.next()
+            relationships.relationshipVisit(relId, relVisitor)
+
+            val relationshipIsUniqueInPath = !rels.contains(relId)
+            if (relationshipIsUniqueInPath) {
+              stack.push((otherSide, rels :+ relId))
+            }
+          }
+        }
+        val needsFlipping = if (dir == SemanticDirection.BOTH)
+          projectedDir == SemanticDirection.INCOMING
+        else
+          dir != projectedDir
+
+        val projectedRels = if (needsFlipping)
+          rels.reverse
+        else
+          rels
+
+        (fromNode, projectedRels)
+      }
+
+      override def hasNext: Boolean = stack.nonEmpty
+    }
+  }
+
+  protected def internalCreateResults(input: Iterator[ExecutionContext], state: QueryState): Iterator[ExecutionContext] = {
+    input.flatMap {
+      inputRowWithFromNode =>
+        val fromNode = inputRowWithFromNode.getLongAt(fromOffset)
+        val paths: Iterator[(LNode, Seq[LRelationship])] = varLengthExpand(fromNode, state, inputRowWithFromNode)
+        paths collect {
+          case (toNode: LNode, relIds: Seq[LRelationship])
+            if relIds.length >= min && isToNodeValid(inputRowWithFromNode, toNode) =>
+            val resultRow = PrimitiveExecutionContext(pipeline)
+            resultRow.copyFrom(inputRowWithFromNode)
+            resultRow.setLongAt(toOffset, toNode)
+            val rels = relIds.map(state.query.relationshipOps.getById)
```

Comment:
Not great to have to hit the store to re-lookup the relationships again, when we know the previous lookup already got all this information.